Modules/_heapqmodule.c

```diff
@@ -538,6 +538,460 @@ _heapq__heapify_max(PyObject *module, PyObject *heap)
+                        arr[pos + 1] = NULL;
+                        return -1;
+                    }
+                    if (cmp > 0) {
+                        childpos += 2;
+                        childkey = otherchildkey;
+                        childvalue = otherchildvalue;
+                    }
+                }
+            }
+            arr[pos] = childkey;
+            arr[pos + 1] = childvalue;
+            pos = childpos;
+        }
+    }
+    else {
+        while (pos < 2 * (n - 1)) {
+            childpos = 2 * pos + 2;
+            childkey = arr[childpos];
+            childvalue = arr[childpos + 1];
+            otherchildkey = arr[childpos + 2];
+            otherchildvalue = arr[childpos + 3];
+            if (otherchildkey != NULL) {
+                if (childkey == NULL) {
+                    childpos += 2;
+                    childkey = otherchildkey;
+                    childvalue = otherchildvalue;
+                }
+                else {
+                    cmp = PyObject_RichCompareBool(
+                        otherchildkey, childkey, Py_LT);
+                    if (cmp < 0) {
+                        arr[pos] = NULL;
+                        arr[pos + 1] = NULL;
+                        return -1;
+                    }
+                    if (cmp > 0) {
+                        childpos += 2;
+                        childkey = otherchildkey;
+                        childvalue = otherchildvalue;
+                    }
+                }
+            }
+            arr[pos] = childkey;
+            arr[pos + 1] = childvalue;
+            pos = childpos;
+        }
+    }
+    arr[pos] = NULL;
+    arr[pos + 1] = NULL;
+
+    it = PyList_GET_ITEM(mo->iterators, pos/2 - (n - 1));
+    if (it == NULL)
+        return 0;
+    childvalue = PyIter_Next(it);
+    if (childvalue == NULL) {
+        Py_DECREF(it);
+        PyList_SET_ITEM(mo->iterators, pos/2 - (n - 1), NULL);
+        if (PyErr_Occurred())
+            return -1;
+    }
+    else {
+        childkey = _PyObject_CallOneArg(mo->keyfunc, childvalue);
+        if (childkey == NULL) {
+            Py_DECREF(childvalue);
+            return -1;
+        }
+        arr[pos] = childkey;
+        arr[pos+1] = childvalue;
+    }
+    return 0;
+}
+
+/* shift = 2 * n - (1 << n.bit_length()) */
+static Py_ssize_t
+get_iter_shift(Py_ssize_t n) {
+    int n0 = n;
+    int m = 1;
+    while (n > 0) {
+        n >>= 1;
+        m <<= 1;
+    }
+    return (2 * n0) - m;
+}
+
+static PyObject *
+merge_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *keyfunc = NULL;
+    int reverse = 0;
+    mergeobject *mo;
+    PyObject *iterators = NULL, *tree = NULL;
+    Py_ssize_t num_iters, shift, i;
+
+    if (kwds != NULL) {
+        char *kwlist[] = {"key", "reverse", 0};
+        PyObject *tmpargs = PyTuple_New(0);
+        if (tmpargs == NULL)
+            return NULL;
+        if (!PyArg_ParseTupleAndKeywords(tmpargs, kwds, "|Op:merge",
```

Comment:
You can use the argument clinic to simplify this code. Please, check out https://docs.python.org/3/howto/clinic.html

Reply:
Is there any documentation on using the argument clinic for passing `*args`-like variable numbers of arguments? The signature should be `_heapq.merge.__new__(*iterables, key=None, reverse=False)`.

Reply:
> Is there any documentation on using the argument clinic for passing `*args`-like variable numbers of arguments? The signature should be `_heapq.merge.__new__(*iterables, key=None, reverse=False)`.

Oh, in that case, that won't be possible :( (I was under the assumption that https://bugs.python.org/issue20291 was resolved already). Apologies :)

Reply:
> being the class private
Maybe I'm confused, but the class should be able to be imported from the `heapq` module, which has an existing API. However, I had in mind that the struct members (e.g. `tree`) would be private to the class.

Reply:
> Maybe I'm confused, but the class should be able to be imported from the `heapq` module,

Nop, my bad. For some reason, I was under the assumption that this was a private helper instead of the accelerator. Disregard this comment (and I just realized that another observation I had was also based on this incorrect assumption).